src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE 1
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <stdlib.h>
+#include <sys/un.h>
+#include <errno.h>
+#include <err.h>
+
+#define N_ELEMENTS(arr) (sizeof (arr) / sizeof ((arr)[0]))
+#define SD_LISTEN_FDS_START 3
+#define MAX_COCKPIT_WS_ARGS 5
+
+static struct instance_type
+{
+  const char *sockname;
+  const char *argv[MAX_COCKPIT_WS_ARGS];
+} instance_types[] = {
+  {"https.sock", {"--for-tls-proxy", "--port", "0"}},
+  {"http-redirect.sock", {"--proxy-tls-redirect", "--no-tls", "--port", "0"}},
+  {"http.sock", {"--no-tls", "--port", "0"}},
+};
+
+static int socket_to_pid[N_ELEMENTS (instance_types)];
+static struct pollfd ufds[N_ELEMENTS (instance_types)];
```

Comment:
Could you expand the "u" here to make this easier to understand? Also, why not simply add an int and a pollfd to `struct instance_type` and rename that to just `instances`? This would make handling the array more robust.

Reply:
Yes for the socket_to_pid. 
But for the ufds (renamed to unixfds) we actually need an array of these structs for the argument of poll 
See ```int poll(struct pollfd *fds, nfds_t nfds, int timeout);```

So I can move each pollfd struct inside the ```struct instance_type``` but then I will need to create an array again for parameter of poll anyway. @martinpitt wdyt?

Reply:
Ah right, if you need the array for poll, then it's fine to keep it of course.